liquidity-amo/contracts/V3AMO.sol

```diff
@@ -0,0 +1,578 @@
+                boost > usd, // zeroForOne
+                type(int256).max,
+                targetSqrtPriceX96
+            );
+            (, int256 usdDelta) = sortAmounts(amount0, amount1);
+            amountIn = uint256(usdDelta);
+        } else if (poolType == PoolType.CL) {
+            IVeloQuoterV2.QuoteExactOutputSingleParams memory params = IVeloQuoterV2.QuoteExactOutputSingleParams({
+                tokenIn: usd,
+                tokenOut: boost,
+                amount: uint256(type(int256).max),
+                tickSpacing: IUniswapV3Pool(pool).tickSpacing(),
+                sqrtPriceLimitX96: targetSqrtPriceX96
+            });
+            (amountIn, , , ) = IVeloQuoterV2(quoter).quoteExactOutputSingle(params);
+        } else if (poolType == PoolType.ALGEBRA_V1_0 || poolType == PoolType.ALGEBRA_V1_9) {
+            (amountIn, ) = IAlgebraQuoter(quoter).quoteExactOutputSingle(
+                usd,
+                boost,
+                uint256(type(int256).max),
+                targetSqrtPriceX96
+            );
+        } else if (poolType == PoolType.ALGEBRA_INTEGRAL) {
+            IAlgebraQuoter.QuoteExactOutputSingleParams memory params = IAlgebraQuoter.QuoteExactOutputSingleParams({
+                tokenIn: usd,
+                tokenOut: boost,
+                amount: uint256(type(int256).max),
+                limitSqrtPrice: targetSqrtPriceX96
+            });
+            (, amountIn, , , , ) = IAlgebraQuoter(quoter).quoteExactOutputSingle(params);
+        } else {
+            IQuoterV2.QuoteExactOutputSingleParams memory params = IQuoterV2.QuoteExactOutputSingleParams({
+                tokenIn: usd,
+                tokenOut: boost,
+                amount: uint256(type(int256).max),
+                fee: IUniswapV3Pool(pool).fee(),
+                sqrtPriceLimitX96: targetSqrtPriceX96
+            });
+            (amountIn, , , ) = IQuoterV2(quoter).quoteExactOutputSingle(params);
+        }
+        liquidity = _getLiquidityForUsdAmount(amountIn);
+        if (liquidity > positionLiquidity) liquidity = positionLiquidity;
+
+        _unfarmBuyBurn(
+            liquidity,
+            1, // minBoostRemove
+            1 // minUsdRemove
+        );
+
+        newBoostPrice = boostPrice();
+    }
+
+    function _validateSwap(bool boostForUsd) internal view override {}
+
+    function _getSqrtPriceX96() internal view returns (uint160 _sqrtPriceX96) {
+        if (poolType == PoolType.SOLIDLY_V3) {
+            (_sqrtPriceX96, , , ) = ISolidlyV3Pool(pool).slot0();
+        } else if (poolType == PoolType.CL) {
+            (_sqrtPriceX96, , , , , ) = ICLPool(pool).slot0();
+        } else if (poolType == PoolType.ALGEBRA_V1_0) {
+            (_sqrtPriceX96, , , , , , ) = IAlgebraV10Pool(pool).globalState();
+        } else if (poolType == PoolType.ALGEBRA_V1_9) {
+            (_sqrtPriceX96, , , , , , , ) = IAlgebraV19Pool(pool).globalState();
```

Comment:
There are only 7 fields in Algebra1.9 globalState. Here is assumes there are 8.

https://github.com/cryptoalgebra/AlgebraV1.9/blob/main/src/core/contracts/base/PoolState.sol#L8

```solidity
  struct GlobalState {
    uint160 price; // The square root of the current price in Q64.96 format
    int24 tick; // The current tick
    uint16 fee; // The current fee in hundredths of a bip, i.e. 1e-6
    uint16 timepointIndex; // The index of the last written timepoint
    uint8 communityFeeToken0; // The community fee represented as a percent of all collected fee in thousandths (1e-3)
    uint8 communityFeeToken1;
    bool unlocked; // True if the contract is unlocked, otherwise - false
  }
```

Reply:
Actually, there are 3 versions (branches) for Algebra1.9: **main**, **ve3.3**, and **direction-based-fee**
And each of them has a different structure for `GlobalState`.
I'm not sure if we need to support all of them or just one. We'll discuss it and give you the final decision.

Reply:
Since all of the V3 forks use a function with the same name (`globalState()` for Algebra and `slot0()` for other forks), and all of these functions return a `uint160` as the first return value for the price, we decided to use `staticcall` and then decode the data manually to get the price value:
```solidity
function _getSqrtPriceX96() internal view returns (uint160 _sqrtPriceX96) {
        bytes memory data;
        if (
            poolType == PoolType.ALGEBRA_V1_0 ||
            poolType == PoolType.ALGEBRA_V1_9 ||
            poolType == PoolType.ALGEBRA_INTEGRAL
        ) {
            (, data) = pool.staticcall(abi.encodeWithSignature("globalState()"));
        } else {
            (, data) = pool.staticcall(abi.encodeWithSignature("slot0()"));
        }
        _sqrtPriceX96 = abi.decode(data, (uint160));
    }
```